lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+
+    * `:precisions_cholesky` - Cholesky decomposition of the precision matrices
+  	(inverses of covariances). This is useful for the model inference.
+
+  ## Examples
+
+      iex> key = Nx.Random.key(12)
+      iex> x = Nx.tensor([[1, 2], [1, 4], [1, 0], [10, 2], [10, 4], [10, 0]])
+      iex> Scholar.Cluster.GaussianMixture.fit(x, num_gaussians: 2, key: key).means
+      Nx.tensor(
+        [
+          [1.0, 2.0],
+          [10.0, 2.0]
+        ]
+      )
+  """
+  deftransform fit(x, opts \\ []) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            """
+            expected input tensor to have shape {n_samples, n_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}\
+            """
+    end
+
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    num_samples = Nx.axis_size(x, 0)
+
+    unless opts[:num_gaussians] <= num_samples do
+      raise ArgumentError,
+            """
+            invalid value for :num_gaussians option: \
+            expected positive integer between 1 and #{inspect(num_samples)}, \
+            got: #{inspect(opts[:num_gaussians])}\
+            """
+    end
+
+    key = Keyword.get_lazy(opts, :key, fn -> Nx.Random.key(System.system_time()) end)
+    fit_n(x, key, opts)
+  end
+
+  defnp fit_n(x, key, opts) do
+    x = to_float(x)
+    num_gaussians = opts[:num_gaussians]
+    num_runs = opts[:num_runs]
+    max_iter = opts[:max_iter]
+    tol = opts[:tol]
+    reg_covar = opts[:reg_covar]
+
+    {best_params, _} =
+      while {
+              best_params =
+                init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar),
+              {
+                run = 1,
+                max_lower_bound = Nx.Constants.neg_infinity(),
+                x,
+                key,
+                max_iter,
+                tol,
+                reg_covar
+              }
+            },
+            run <= num_runs do
+        {{lower_bound, params}, _} =
+          while {
+                  {
+                    lower_bound = Nx.Constants.neg_infinity(),
+                    params =
+                      init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar)
+                  },
+                  {
+                    iter = 1,
+                    converged = Nx.tensor(false),
+                    x,
+                    key,
+                    max_iter,
+                    tol,
+                    reg_covar
+                  }
+                },
+                iter <= max_iter and not converged do
+            prev_lower_bound = lower_bound
+            {weights, means, _, precisions_cholesky} = params
+            {log_prob_norm, log_resp} = e_step(x, weights, means, precisions_cholesky)
+            params = m_step(x, log_resp, reg_covar)
+            lower_bound = log_prob_norm
+            change = lower_bound - prev_lower_bound
+            converged = Nx.abs(change) < tol
```

Comment:
```suggestion
            converged = iter > 1 and Nx.abs(change) < tol
```

Reply:
Why is `iter > 1` necessary here?

Reply:
My previous version of the comment had an explanation haha

I think it would be nice to avoid relying on comparison with neg infinity (the initial value) in the first iteration

Reply:
Oh, good catch! But it's probably still not needed.
`Nx.subtract(12, Nx.Constants.neg_infinity) |> Nx.abs`
gives
```
#Nx.Tensor<
  f32
  Inf
>
```
which is larger than any `tol`. :)

Reply:
Yeah, I just don't _like_ comparisons and operations with inf. But even in the worst case where we get `lower_bound = neg_infinity`, `neg_infinity - neg_infinity = nan`, which will also fail your check because it breaks on `true`

Reply:
The only way `lower_bound` ends up being `neg_infinity` at that point is due to numerical error as it's updated `lower_bound = log_prob_norm` at line 169. But yeah, let's avoid operations with infinity.